test/Common/MessageSorterTests.cs

```diff
@@ -0,0 +1,316 @@
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public void WackySystemClock()
+        {
+            var senderId = "A";
+            var receiverId = "B";
+
+            var senderSorter = new MessageSorter();
+
+            // simulate system clock that goes backwards - mechanism should still guarantee monotonicitty
+            var message1 = Send(senderId, receiverId, "1", senderSorter, DateTime.UtcNow);
+            var message2 = Send(senderId, receiverId, "2", senderSorter, DateTime.UtcNow - TimeSpan.FromSeconds(1));
+            var message3 = Send(senderId, receiverId, "3", senderSorter, DateTime.UtcNow - TimeSpan.FromSeconds(2));
+
+            List<RequestMessage> batch;
+            MessageSorter receiverSorter = new MessageSorter();
+
+            // delivering the sequence in order produces 1 message each time
+            receiverSorter = new MessageSorter();
+            batch = receiverSorter.ReceiveInOrder(message1, ReorderWindow).ToList();
+            Assert.Single(batch).Input.Equals("1");
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
+            Assert.Single(batch).Input.Equals("2");
+            batch = receiverSorter.ReceiveInOrder(message3, ReorderWindow).ToList();
+            Assert.Single(batch).Input.Equals("3");
+
+            Assert.Equal(0, receiverSorter.NumberBufferedRequests);
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public void DelayedElement()
+        {
+            var senderId = "A";
+            var receiverId = "B";
+
+            var senderSorter = new MessageSorter();
+
+            var message1 = Send(senderId, receiverId, "1", senderSorter, DateTime.UtcNow);
+            var message2 = Send(senderId, receiverId, "2", senderSorter, DateTime.UtcNow);
+            var message3 = Send(senderId, receiverId, "3", senderSorter, DateTime.UtcNow);
+
+            List<RequestMessage> batch;
+            MessageSorter receiverSorter;
+
+            // delivering first message last delays all messages until getting the first one
+            receiverSorter = new MessageSorter();
+            batch = receiverSorter.ReceiveInOrder(message2, ReorderWindow).ToList();
+            Assert.Empty(batch);
+            batch = receiverSorter.ReceiveInOrder(message3, ReorderWindow).ToList();
+            Assert.Empty(batch);
+            batch = receiverSorter.ReceiveInOrder(message1, ReorderWindow).ToList();
+            Assert.Collection(
+                batch,
+                first => first.Input.Equals("1"),
+                second => second.Input.Equals("2"),
+                third => third.Input.Equals("3"));
+
+            Assert.Equal(0, receiverSorter.NumberBufferedRequests);
+        }
+
+        [Fact]
+        [Trait("Category", PlatformSpecificHelpers.TestCategory)]
+        public void Collection()
```

Comment:
The name of this test isn't very clear to me. It seems like it is very similar to `SimpleInOrder`, just with an extra delay of `ReorderWindows` for `message2` and `message3`. Why is that extra delay important for this scenario?

Reply:
Yeah, this test is not quite doing what it is supposed to do. It should check that duplicates are not filtered or reordered when we get past the `ReorderWindow`. I will update the test name, code, and comments to better reflect the intended scenario.